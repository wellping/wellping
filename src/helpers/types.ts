import {
  AnswerData,
  AnswersList,
} from "@wellping/study-schemas/lib/answerTypes";
import { Question, QuestionsList } from "@wellping/study-schemas/lib/types";

export type DataValidationFunction = () => boolean;

export interface QuestionScreenProps {
  question: Question;
  loadingCompleted: () => void;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: DataValidationFunction) => void;
}
