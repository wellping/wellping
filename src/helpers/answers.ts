import { AnswerData, Answer } from "wellping-study-file/lib/answerTypes";
import { AnswerSchema } from "wellping-study-file/lib/schemas/Answer";
import { Question, Ping } from "wellping-study-file/lib/types";

import {
  getAnswersPingIdsQuestionIdsListAsync,
  clearAnswersPingIdsQuestionIdsListAsync,
} from "./asyncStorage/answersPingIdsQuestionIdsList";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
  secureRemoveAnswerAsync,
} from "./secureStore/answer";

export async function getAnswersAsync({
  unuploadedOnly = false,
}: {
  unuploadedOnly?: boolean;
} = {}): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync({
    unuploadedOnly,
  });

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}

export async function clearAllAnswersAsync() {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();
  await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      await secureRemoveAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
    }),
  );
  await clearAnswersPingIdsQuestionIdsListAsync();
}
