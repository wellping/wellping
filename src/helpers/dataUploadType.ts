import { Answer } from "./answerTypes";
import { UserInstallationInfo } from "./debug";
import { Ping } from "./types";

export type UserData = {
  username: string;
  loginSessionId: string;
  installation: UserInstallationInfo;
};

export interface UploadData {
  user: UserData;
}
export interface AllData extends UploadData {
  pings: Ping[];
  answers: Answer[];
}
export interface UnuploadedData extends UploadData {
  unuploadedPings: Ping[];
  unuploadedAnswers: Answer[];
}

export enum UploadDataType {
  ALL_DATA,
  UNUPLOADED_DATA,
}
export function getUploadDataType(uploadData: UploadData): UploadDataType {
  if ("unuploadedAnswers" in uploadData) {
    return UploadDataType.UNUPLOADED_DATA;
  } else {
    return UploadDataType.ALL_DATA;
  }
}
