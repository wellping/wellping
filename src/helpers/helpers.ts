import {
  QuestionId,
  PlaceholderReplacementValueTreatmentOptions,
  StudyInfo,
} from "@wellping/study-file/lib/types";
import { addMinutes } from "date-fns";
import * as Crypto from "expo-crypto";

export {
  QuestionType,
  QUESTION_TYPES,
  NON_USER_QUESTION_TYPES,
} from "@wellping/study-file/lib/helpers";

// This is mostly for Beiwe.
export async function getHashedPasswordAsync(
  password: string,
): Promise<string> {
  return await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    password,
    {
      encoding: Crypto.CryptoEncoding.BASE64,
    },
  );
}

// This is mostly for Beiwe.
export function base64ToBase64URL(input: string): string {
  return input.replace(/\+/g, "-").replace(/\//g, "_");
}

export const decapitalizeFirstCharacter = (s: string): string => {
  return s.charAt(0).toLowerCase() + s.slice(1);
};

export const withVariable = (varName: string) => `[__${varName}__]`;

export const withPreviousAnswer = (questionId: QuestionId) =>
  `{PREV:${questionId}}`;

export const replacePreviousAnswerPlaceholdersWithActualContent = (
  s: string,
  getAnswerFromQuestionId: (questionId: QuestionId) => string | null,
) => {
  let output = s;

  // https://stackoverflow.com/a/6323598/2603230
  const re = /\{PREV:(.*?)\}/g;
  let m: RegExpExecArray;

  do {
    m = re.exec(s)!;
    if (m) {
      const prevAnswer = getAnswerFromQuestionId(m[1]);
      if (prevAnswer) {
        // https://stackoverflow.com/a/56136657/2603230
        output = output.split(m[0]).join(prevAnswer);
      }
    }
  } while (m);

  return output;
};

const _treatPlaceholderReplacementValue = (
  value: string,
  treatment?: PlaceholderReplacementValueTreatmentOptions,
): string => {
  if (treatment === undefined) {
    return value;
  }

  let treatedValue = value;

  let shouldDecap = false;
  const decapOptions = treatment.decapitalizeFirstCharacter;
  if (decapOptions && decapOptions.enabled) {
    if (decapOptions.includes) {
      if (decapOptions.includes.includes(value)) {
        shouldDecap = true;
      }
    } else if (decapOptions.excludes) {
      if (!decapOptions.excludes.includes(value)) {
        shouldDecap = true;
      }
    } else {
      shouldDecap = true;
    }
  }

  if (shouldDecap) {
    treatedValue = decapitalizeFirstCharacter(treatedValue);
  }

  return treatedValue;
};

export const treatPlaceholderReplacementValue = (
  key: string,
  value: string,
  studyInfo: StudyInfo,
): string => {
  return _treatPlaceholderReplacementValue(
    value,
    // https://stackoverflow.com/a/58780897/2603230
    studyInfo.specialVariablePlaceholderTreatments?.[key],
  );
};

// https://stackoverflow.com/a/2450976/2603230
export function shuffle(array: any[]): any[] {
  let currentIndex = array.length,
    temporaryValue: any,
    randomIndex: number;

  // While there remain elements to shuffle...
  while (currentIndex !== 0) {
    // Pick a remaining element...
    randomIndex = Math.floor(Math.random() * currentIndex);
    currentIndex -= 1;

    // And swap it with the current element.
    temporaryValue = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temporaryValue;
  }

  return array;
}

/**
 * Use local time zone for date. In other words, ignore the timezone suffix of
 * the date.
 */
export function ignoreTimeObjectTimezone(date: Date): Date {
  return addMinutes(date, new Date().getTimezoneOffset());
}
