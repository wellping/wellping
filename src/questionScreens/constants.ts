export { SLIDER_DEFAULTS } from "@wellping/study-file/lib/schemas/Question";
