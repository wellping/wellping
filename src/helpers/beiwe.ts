/**
 * Unless otherwise noted, precondition for every function in the file is:
 * `useBeiwe(studyInfo) === true`.
 */

import * as Device from "expo-device";
import { Platform } from "react-native";

import { UploadData } from "./dataUpload";
import { JS_VERSION_NUMBER } from "./debug";
import { base64ToBase64URL, getHashedPasswordAsync } from "./helpers";
import { getLoginSessionIDAsync } from "./loginSession";
import { User, secureGetUserAsync } from "./secureStore/user";
import { DataUploadServerResponse, getBeiweServerConfig } from "./server";
import { getStudyInfoAsync } from "./studyFile";

/**
 * If success, returns nothing.
 * If error, throws error message.
 */
export async function beiweLoginAsync(user: User): Promise<void> {
  const request: { [key: string]: any } = {
    new_password: user.password, // Do not reset password
    phone_number: "1234567890", // Doesn't matter
    device_os: Device.osName,
    os_version: Device.osVersion,
    product: Device.modelName,
    brand: Device.brand,
    manufacturer: Device.manufacturer,
    model: Device.modelId,
    beiwe_version: `WellPing${JS_VERSION_NUMBER}`,
  };

  try {
    let endpoint = "/register_user";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    await makePostRequestAsync({
      endpoint,
      request,
      body: {},
      user,
    });
  } catch (e) {
    throw new Error(`Request error: ${e}.`);
  }
}

/**
 * Returns a `DataUploadServerResponse` if successful.
 * Throws an error otherwise.
 */
export async function beiweUploadDataForUserAsync(
  data: UploadData,
  user: User,
  startUploading: () => void,
  endUploading: (errorMessage?: string) => void,
): Promise<DataUploadServerResponse> {
  startUploading();

  try {
    const endpoint = "/ssnl_upload";
    const response = await makePostRequestAsync({
      endpoint,
      request: {},
      body: data,
      user,
      throwShortError: true,
    });
    //console.log(`${JSON.stringify(response)}`);
    endUploading();
    return response as DataUploadServerResponse;
  } catch (e) {
    endUploading(`BW: ${e.message}`);
    throw e;
  }
}

async function getRequestURLAsync(
  endpoint: string,
  request: { [key: string]: any } = {},
  forUser?: User,
): Promise<string> {
  let user: User | null = forUser || null;
  if (!user) {
    user = await secureGetUserAsync();

    if (user == null) {
      throw new Error("user == null in getRequestURLAsync");
    }
  }

  const passwordHash = await getHashedPasswordAsync(user.password);
  request["password"] = base64ToBase64URL(passwordHash);
  request["device_id"] = await getLoginSessionIDAsync(user);
  request["username"] = user.username;
  request["patient_id"] = user.username;

  //console.warn(`request is ${JSON.stringify(request)}`);

  const studyInfo = await getStudyInfoAsync();
  const serverUrl = getBeiweServerConfig(studyInfo).serverUrl;
  const query = Object.keys(request)
    .map((k) => encodeURIComponent(k) + "=" + encodeURIComponent(request[k]))
    .join("&");
  const url = `${serverUrl}${endpoint}?${query}`;
  return url;
}

async function makePostRequestAsync({
  endpoint,
  request,
  body,
  user,
  throwShortError,
}: {
  endpoint: string;
  request: { [key: string]: any };
  body?: { [key: string]: any };
  user?: User;
  throwShortError?: boolean;
}): Promise<any> {
  const headers = {
    "Beiwe-Api-Version": "2",
    "WellPing-JSVersion": JS_VERSION_NUMBER,
    Accept: "*/*",
    "Content-Type": "application/json",
  };

  const url = await getRequestURLAsync(endpoint, request, user);

  //console.warn("url is " + url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  if (response.status < 200 || response.status >= 400) {
    if (response.status === 401 || response.status === 403) {
      throw new Error(
        throwShortError
          ? `VF${response.status}`
          : `Verification failed.\n\nRequest: ${JSON.stringify(
              request,
            )}.\n\nResponse: ${JSON.stringify(response)}.`,
      );
    }

    throw new Error(
      throwShortError
        ? `OF${response.status}`
        : `Request not successful.\n\nRequest: ${JSON.stringify(
            request,
          )}.\n\nResponse: ${JSON.stringify(response)}.`,
    );
  }

  //console.warn(response);

  return response.json();
}
