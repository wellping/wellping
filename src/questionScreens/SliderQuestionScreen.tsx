import Slider from "@react-native-community/slider";
import {
  SliderAnswerData,
  SliderAnswer,
} from "@wellping/study-schemas/lib/answerTypes";
import { SliderQuestion } from "@wellping/study-schemas/lib/types";
import React from "react";
import { View, Text } from "react-native";

import {
  getNonCriticalProblemTextForUser,
  alertWithShareButtonContainingDebugInfoAsync,
} from "../helpers/debug";
import { QuestionScreenProps } from "../helpers/types";
import { SLIDER_DEFAULTS } from "./constants";

export const getQuestionDefaultSliderValue = (question: SliderQuestion) => {
  if (question.defaultValue == null) {
    return SLIDER_DEFAULTS.DEFAULT_VALUE;
  }
  return question.defaultValue;
};

interface SliderQuestionScreenProps extends QuestionScreenProps {
  question: SliderQuestion;
}

const SliderQuestionScreen: React.ElementType<SliderQuestionScreenProps> = ({
  question,
  loadingCompleted,
  onDataChange,
  allAnswers,
  allQuestions,
  pipeInExtraMetaData,
}) => {
  React.useEffect(() => {
    loadingCompleted();
  }, []);

  let defaultSliderValue = getQuestionDefaultSliderValue(question);
  if (question.defaultValueFromQuestionId) {
    const prevQuestionAnswer = allAnswers[
      pipeInExtraMetaData(question.defaultValueFromQuestionId)
    ] as SliderAnswer;
    if (prevQuestionAnswer && prevQuestionAnswer.data != null) {
      defaultSliderValue = prevQuestionAnswer.data.value;
    } else {
      // If the user did not answer the previous question,
      // use the default slider value from that question instead.
      const prevQuestion = allQuestions[
        question.defaultValueFromQuestionId
      ] as SliderQuestion;
      if (prevQuestion == null) {
        alertWithShareButtonContainingDebugInfoAsync(
          getNonCriticalProblemTextForUser(
            `defaultValueFromQuestionId ${question.defaultValueFromQuestionId} prevQuestion == null`,
          ),
        );
      } else {
        defaultSliderValue = getQuestionDefaultSliderValue(prevQuestion);
      }
    }
  }

  // This value is for display only - it should not be used for anything else.
  const [__FOR_DISPLAY_ONLY__sliderValue, __FOR_DISPLAY_ONLY__setSliderValue] =
    React.useState<number | null>(null);

  return (
    <View style={{ paddingVertical: 30 }}>
      <Slider
        step={question.step ?? SLIDER_DEFAULTS.STEP}
        value={defaultSliderValue}
        minimumValue={question.minimumValue ?? SLIDER_DEFAULTS.MIN_VALUE}
        maximumValue={question.maximumValue ?? SLIDER_DEFAULTS.MAX_VALUE}
        minimumTrackTintColor="#2F2424"
        maximumTrackTintColor="#2F2424"
        onValueChange={(value) => {
          if (question.displayCurrentValueToUser) {
            // We need to set it only if it is displayed.
            __FOR_DISPLAY_ONLY__setSliderValue(value);
          }
        }}
        onSlidingComplete={(value) => {
          onDataChange({ value } as SliderAnswerData);
        }}
        accessibilityLabel="slider input"
      />
      <View style={{ flexDirection: "row", justifyContent: "space-between" }}>
        <Text style={{ maxWidth: "40%" }}>{question.slider[0]}</Text>
        <Text style={{ maxWidth: "40%" }}>{question.slider[1]}</Text>
      </View>
      {question.displayCurrentValueToUser &&
        __FOR_DISPLAY_ONLY__sliderValue !== null && (
          <Text style={{ marginTop: 20, textAlign: "center" }}>
            (You have selected {__FOR_DISPLAY_ONLY__sliderValue})
          </Text>
        )}
    </View>
  );
};

export default SliderQuestionScreen;
