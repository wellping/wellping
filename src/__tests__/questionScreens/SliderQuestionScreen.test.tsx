import { SliderAnswerData } from "@wellping/study-schemas/lib/answerTypes";
import { SliderQuestion } from "@wellping/study-schemas/lib/types";
import React from "react";
import { render } from "react-native-testing-library";

import { QuestionType } from "../../helpers/helpers";
import SliderQuestionScreen, {
  getQuestionDefaultSliderValue,
} from "../../questionScreens/SliderQuestionScreen";
import { SLIDER_DEFAULTS } from "../../questionScreens/constants";
import { simplePipeInExtraMetaData } from "../helper";
import {
  moveSlider,
  findSliderAsync,
} from "../reactNativeTestingLibraryHelper";

const SLIDER_QUESTIONS: {
  [questionId: string]: SliderQuestion;
} = {
  WithoutDefault: {
    id: "WithoutDefault",
    type: QuestionType.Slider,
    question: "This question does not have any default value settings.",
    slider: ["leftmost", "rightmost"],
    next: "WithConstantDefault",
  },
  WithConstantDefault: {
    id: "WithConstantDefault",
    type: QuestionType.Slider,
    question: "This question has a constant default value.",
    defaultValue: 25,
    slider: ["left", "right"],
    next: "WithDefaultFromElsewhere",
  },
  WithDefaultFromQWithoutDefault: {
    id: "WithDefaultFromQWithoutDefault",
    type: QuestionType.Slider,
    question:
      "This question has a default value that is from another question (that does not have `defaultValue`).",
    defaultValueFromQuestionId: "WithoutDefault",
    defaultValue: 100, // This should be ignored for this question.
    slider: ["left", "right"],
    next: "WithDefaultFromQWithDefault",
  },
  WithDefaultFromQWithDefault: {
    id: "WithDefaultFromQWithDefault",
    type: QuestionType.Slider,
    question:
      "This question has a default value that is from another question (that has `defaultValue`).",
    defaultValueFromQuestionId: "WithDefaultFromQWithoutDefault",
    defaultValue: 0, // This should be ignored for this question.
    slider: ["left", "right"],
    next: null,
  },
};
const WITHOUT_DEFAULT = SLIDER_QUESTIONS["WithoutDefault"];
const WITH_CONSTANT_DEFAULT = SLIDER_QUESTIONS["WithConstantDefault"];
const WITH_DEFAULT_FROM_Q_WITHOUT_DEFAULT =
  SLIDER_QUESTIONS["WithDefaultFromQWithoutDefault"];
const WITH_DEFAULT_FROM_Q_WITH_DEFAULT =
  SLIDER_QUESTIONS["WithDefaultFromQWithDefault"];

test("without default value", async () => {
  const mockLoadingCompleted = jest.fn();
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const renderResults = render(
    <SliderQuestionScreen
      key={WITHOUT_DEFAULT.id}
      question={WITHOUT_DEFAULT}
      loadingCompleted={mockLoadingCompleted}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );
  const { toJSON } = renderResults;

  const sliderInput = await findSliderAsync(renderResults);
  expect(sliderInput.props.value).toBe(SLIDER_DEFAULTS.DEFAULT_VALUE);

  expect(mockLoadingCompleted).toHaveBeenCalledTimes(1);

  // Because there isn't a need to pipe in any data.
  expect(mockPipeInExtraMetaData).not.toHaveBeenCalled();

  // Because there isn't any requirement to validate the data.
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  expect(toJSON()).toMatchSnapshot();
});

test("with constant default value", async () => {
  const mockLoadingCompleted = jest.fn();
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const renderResults = render(
    <SliderQuestionScreen
      key={WITH_CONSTANT_DEFAULT.id}
      question={WITH_CONSTANT_DEFAULT}
      loadingCompleted={mockLoadingCompleted}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );
  const { toJSON } = renderResults;

  const sliderInput = await findSliderAsync(renderResults);
  expect(sliderInput.props.value).toBe(WITH_CONSTANT_DEFAULT.defaultValue);

  expect(mockLoadingCompleted).toHaveBeenCalledTimes(1);

  // Because there isn't a need to pipe in any data.
  expect(mockPipeInExtraMetaData).not.toHaveBeenCalled();

  // Because there isn't any requirement to validate the data.
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  expect(toJSON()).toMatchSnapshot();
});

test.each([
  ["without default value", { value: 31 }, WITH_DEFAULT_FROM_Q_WITHOUT_DEFAULT],
  ["without default value", null, WITH_DEFAULT_FROM_Q_WITHOUT_DEFAULT],
  ["with default value", { value: 61 }, WITH_DEFAULT_FROM_Q_WITH_DEFAULT],
  ["with default value", null, WITH_DEFAULT_FROM_Q_WITH_DEFAULT],
] as [string, SliderAnswerData | null, SliderQuestion][])(
  "with default value from a question %s with data `%o`",
  async (_, prevAnswerData, question) => {
    const mockLoadingCompleted = jest.fn();
    const mockOnDataChangeFn = jest.fn();
    const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
    const mockSetDataValidationFunction = jest.fn();

    const renderResults = render(
      <SliderQuestionScreen
        key={question.id}
        question={question}
        loadingCompleted={mockLoadingCompleted}
        onDataChange={mockOnDataChangeFn}
        allAnswers={{
          // @ts-ignore (we don't need to craft an entire Answer)
          [question.defaultValueFromQuestionId]: {
            data: prevAnswerData,
          },
        }}
        allQuestions={SLIDER_QUESTIONS}
        pipeInExtraMetaData={mockPipeInExtraMetaData}
        setDataValidationFunction={mockSetDataValidationFunction}
      />,
    );
    const { toJSON } = renderResults;

    const sliderInput = await findSliderAsync(renderResults);
    // `defaultValue` should be ignored here.
    expect(sliderInput.props.value).not.toBe(question.defaultValue);

    expect(mockLoadingCompleted).toHaveBeenCalledTimes(1);

    // To pipe in `defaultValueFromQuestionId`.
    expect(mockPipeInExtraMetaData).toHaveBeenCalledTimes(1);

    // Because there isn't any requirement to validate the data.
    expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

    expect(sliderInput.props.value).toBe(
      prevAnswerData
        ? prevAnswerData.value // If previous data is `null`, the question should use the default
        : // slider value of previous question.
          getQuestionDefaultSliderValue(
            SLIDER_QUESTIONS[question.defaultValueFromQuestionId!],
          ),
    );

    expect(toJSON()).toMatchSnapshot();
  },
);

test("update values", async () => {
  const mockLoadingCompleted = jest.fn();
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const renderResults = render(
    <SliderQuestionScreen
      key={WITHOUT_DEFAULT.id}
      question={WITHOUT_DEFAULT}
      loadingCompleted={mockLoadingCompleted}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );

  const sliderInput = await findSliderAsync(renderResults);

  expect(mockLoadingCompleted).toHaveBeenCalledTimes(1);

  const NEW_VALUES = [88, 99, 66, 55];
  for (let i = 0; i < NEW_VALUES.length; i++) {
    const NEW_VALUE = NEW_VALUES[i];
    moveSlider(sliderInput, NEW_VALUE);
    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(i + 1, {
      value: NEW_VALUE,
    } as SliderAnswerData);
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(i + 1);
  }
});
