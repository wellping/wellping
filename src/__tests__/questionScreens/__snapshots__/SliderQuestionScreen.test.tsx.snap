// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with constant default value 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={25}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question with default value with data \`{ value: 61 }\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={61}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question with default value with data \`null\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={100}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question without default value with data \`{ value: 31 }\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={31}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question without default value with data \`null\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={50}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`without default value 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RNCSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    inverted={false}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={[Function]}
    onRNCSliderSlidingComplete={[Function]}
    onRNCSliderSlidingStart={null}
    onRNCSliderValueChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    tapToSeek={false}
    thumbImage={null}
    value={50}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      leftmost
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      rightmost
    </Text>
  </View>
</View>
`;
